_SPIKE_.md

```diff
@@ -0,0 +1,543 @@
+  `columns` field too. This field should contain an array of column-index
+  specification, for example:
+
+  ```js
+  ['email ASC', 'lower(name) DESC'];
+  ```
+
+  In the contrary with the current implementation, I am proposing that `columns`
+  should replace any column list created from `properties` and `keys`, and that
+  `columns` should be nested inside connector-specific options:
+
+  ```js
+  {
+    properties: {
+      // default setup used by most connectors
+      email: 'ASC',
+    },
+    postgresql: {
+      // in PostgreSQL, use a different index definition
+      columns: ['lower(email) ASC']
+    },
+  }
+  ```
+
+### Foreign keys at property level
+
+Introduce a new property metadata "references" (inspired by ANSI SQL):
+
+```js
+{
+  categoryId: {
+    type: 'number',
+    references: {
+      // a TypeResolver
+      model: () => Category,
+      // or a Model class
+      model: Category
+      // or a Model name
+      model: 'Category'
+
+      // name of the target property
+      property: 'id',
+
+      // referential actions
+      onUpdate: 'CASCADE',
+      onDelete: 'CASCADE',
+    }
+  }
+}
+```
+
+Support connector-specific configuration too.
+
+```js
+{
+  categoryId: {
+    type: 'number',
+    mysql: {
+      references: {
+        model: () => Category,
+        property: 'id',
+      }
+    }
+  }
+}
+```
+
+### Foreign keys at model level
+
+Modify the current connector-dependant syntax to make it easier to read and
+support composite foreign keys too.
+
+```js
+{
+  foreignKeys: {
+    [keyName]: {
+      // optional, overrides keyName
+      name: 'constraint_name_for_db',
+
+      // Property name(s) (will be mapped to column name)
+      // formerly: foreignKey
+      sourceProperties: ['source property name'],
+
+
+      // formerly: entity
+      targetModel: 'TargetModel',
+
+      // Property name(s) (will be mapped to column name)
+      // formerly: entityKey
+      targetProperties: ['target property name'],
+
+      // referential actions
+      onUpdate: 'CASCADE',
+      onDelete: 'CASCADE',
+    }
+  }
+}
+```
+
+### Decorators
+
+I am proposing to reuse existing `@property` and `@model` decorators. They are
+accepting full Model/Property definition interfaces, no additional changes are
+needed to make them support the new index/FK syntax.
+
+### Additional changes
+
+When a model specifies an index or constraint that's not supported by the
+connector running automigration, the connector should let the user know about
+the problem. To preserve backwards compatibility, I am proposing to print a
+console warning only.
+
+We can also introduce a new `options` flag for `autoupdate`/`automigrate` to
+control what happens when a model definition specifies metadata that the
+connector cannot process. The flag can allow three values:
+
+- true: abort migration with error
+- 'warn': print a warning and ignore unsupported metadata
```

Comment:
sounds good, I would suggest `warn` as default.

Reply:
+1 on warning to be the default.